<a href="https://colab.research.google.com/github/Nikkuchan/AIML-Problem-Sets/blob/main/58089_PrelimPS_NicoleGala.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Topic02a : Prelim Problem Set I

#**SOLUTIONS FOR CASE 1** 
**[Problem 1.a]** 

In [1]:
import numpy as np
import math
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
x = np.array([[-1, 1, 1/32], [1/2, -2, 0], [-1, 3/7, 4/5]])
x

array([[-1.        ,  1.        ,  0.03125   ],
       [ 0.5       , -2.        ,  0.        ],
       [-1.        ,  0.42857143,  0.8       ]])

$$
x =\begin{bmatrix} -1 & 1 & \frac{1}{32} \\ \frac{1}{2} & {-2} & 0\\ -1 & \frac{3}{7} & \frac{4}{5}\end{bmatrix}
$$

**[Problem 1.b]** 

In [3]:
csc = 1/math.sin(2**180)
cos = math.cos(180)
sin = math.sin(180)

t = np.array([cos, sin, csc]) ##Values from the upper part
t

array([-0.59846007, -0.80115264,  1.00032034])

$$
t =\begin{bmatrix} cos{(\theta)}\hat{i} \\ sin{(\theta)}\hat{j}\\ -csc{(2\theta)}\hat{k}\end{bmatrix}
$$

**[Problem 1.c]** 

In [4]:
attendees = np.array([200, 45, 15, 40])

np.true_divide(attendees,300) * 100

array([66.66666667, 15.        ,  5.        , 13.33333333])

$$
\begin{matrix} {Students} \\ {Professionals} \\ {Panels} \\ {Committee} \end{matrix}
\begin{bmatrix} 200 \\ 45 \\ 15 \\ 40 \end{bmatrix} =
\begin{bmatrix} 66.67\% \\ 15.00\% \\ 5.00\% \\ 13.33\% \end{bmatrix}
$$

#**SOLUTIONS FOR CASE 2** 
**[Problem 2.a]** 

In [6]:
import numpy as np

def vect_mag(a):
	return np.sqrt(sum(k**2 for k in a))

vector = np.array([1, 2, 3, 4, 5])
v = vect_mag(vector)

print (vector, v, sep = '\n')

[1 2 3 4 5]
7.416198487095663


**[Problem 2.b]** 

In [8]:
def angle_vectors(a,b):
  inner = np.inner(a, b)
  norms = np.linalg.norm(a) * np.linalg.norm(b)

  cos = inner / norms
  rad = np.arccos(np.clip(cos, -1.0, 1.0))
  deg = np.rad2deg (rad)
  print("Radiant: ", rad)
  print("Degree: ", deg)
  return deg,rad


In [9]:
  a = np.array([9, 5])
  b = np.array([-6, 2])
  print(a,b)

[9 5] [-6  2]


In [10]:
angle_vectors(a,b)

Radiant:  2.3127435948008137
Degree:  132.51044707800082


(132.51044707800082, 2.3127435948008137)

#**SOLUTIONS FOR CASE 3**
**[Problem 3.a]**
> Unlike scalar multiplication, multiplication involving matrices is not commutative. However, commutative property is allowed when getting the dot product between two vectors.

It means that $A\cdot B = B \cdot A $ can be applied even if one of the vectors is transposed. 

In problem 3.a, it is required to prove the equality of $X \cdot \theta$ with $\theta^T\cdot X$ given that $z$ can also be solved as $X \cdot \theta$ in equation 3.1
$$X = \begin{bmatrix} 
— (x^{(1)})^T— \\ 
— (x^{(2)})^T— \\
\vdots \\
— (x^{(m)})^T— \\
\end{bmatrix} \text{, } 
\theta = \begin{bmatrix} 
\theta^{(1)} \\ 
\theta^{(2)} \\
\vdots \\
\theta^{(m)} \\
\end{bmatrix} $$

By computing the vector product of the dataset and the parameters $$ z = X\cdot \theta$$ we get:

$$z = \begin{bmatrix} 
— (x^{(1)})^T\theta— \\ 
— (x^{(2)})^T\theta-\\
\vdots \\
— (x^{(m)})^T\theta— \\
\end{bmatrix}$$


>By applying commutative property to $ z = X\cdot \theta$ which is the vectorized form of $x^{(i)^T} \theta^{(i)}$ , we can say that it is equal to 
$$ z = \theta^T\cdot X$$ 

$$z = \begin{bmatrix} 
—\theta^T(x^{(1)})— \\ 
—\theta^T(x^{(2)})-\\
\vdots \\
—\theta^T(x^{(m)})— \\
\end{bmatrix}$$

Hence, 
$$z = \begin{bmatrix} 
— (x^{(1)})^T\theta— \\ 
— (x^{(2)})^T\theta-\\
\vdots \\
— (x^{(m)})^T\theta— \\
\end{bmatrix} 
= \begin{bmatrix} 
—\theta^T(x^{(1)})— \\ 
—\theta^T(x^{(2)})-\\
\vdots \\
—\theta^T(x^{(m)})— \\
\end{bmatrix} $$


**[Problem 3.b]**
>The output derived from logic regression hypothesis $h_\theta$ is used in determining the estimated probability. In this problem, the shape of $h_\theta$ must be determined given that $X$ has a shape of $(300,5)$

Logic Regression Hypothesis: 
$$ h_\theta(x) = g(z)\$$

wherein
$$g(z) = \frac{1}{1+e^{-z}}\$$
and

$$ z = X\cdot\theta$$ 

if $X$ has a shape of $(300,5)$ then

$$X = \begin{bmatrix} 
x^{1} \quad x^{301} \quad x^{601} \quad x^{901} \quad x^{1201} \\
x^{2} \quad x^{302} \quad x^{603} \quad x^{904} \quad x^{1205} \\  
\enspace\vdots\hspace{2em} \vdots\hspace{2em}  \vdots\hspace{2em} \vdots \hspace{2em}\vdots\enspace\\
x^{300} \quad x^{600} \quad x^{900} \quad x^{1200} \quad x^{1500} \\  
\end{bmatrix}
\text{, } 
\theta = \begin{bmatrix} 
\theta^{1} \\ 
\theta^{2} \\
\vdots \\
\theta^{k} \\
\end{bmatrix}
$$

It can also be written as 
$$ h_\theta = g(X.\theta)\$$

Getting $h_\theta$ by multiplying $X$ that has a shape of $(n,m)$ or (300,5) to its parameter vector $\theta$ that has a dimension of $(m,k)$ or (300,1) would result to getting the scalar value of the inner product $(300,1)$

$$h_\theta = \begin{bmatrix} 
\theta^{1}(x^{1}) \\ 
\theta^{2}(x^{2}) \\
\vdots \\
\theta^{300}(x^{300}) \\ 
\end{bmatrix}
$$

**[Problem 3.c]**


$$\begin{bmatrix} 
\frac{\partial{J}}{\partial{\theta}_0} \\ 
\frac{\partial{J}}{\partial{\theta}_1} \\
\frac{\partial{J}}{\partial{\theta}_2} \\
\vdots \\
\frac{\partial{J}}{\partial{\theta}_n} \\
\end{bmatrix} = 
\frac{1}{m}
\begin{bmatrix} 
\sum^{m}_{i=1} [ -y^{(i)} \log (h_{\theta} (x^{(i)})) - (1-y^{(i)}) \log (1-h_{\theta}(x^{(i)}))] \\
\sum^{m}_{i=1} [ -y^{(i)} \log (h_{\theta} (x^{(i)})) - (1-y^{(i)}) \log (1-h_{\theta}(x^{(i)}))] \\ 
\sum^{m}_{i=1} [ -y^{(i)} \log (h_{\theta} (x^{(i)})) - (1-y^{(i)}) \log (1-h_{\theta}(x^{(i)}))] \\ 
\vdots \\
\sum^{m}_{i=1} [ -y^{(i)} \log (h_{\theta} (x^{(i)})) - (1-y^{(i)}) \log (1-h_{\theta}(x^{(i)}))] \\ 
\end{bmatrix} 
$$

$$
\frac{\partial{J}(\theta)}{\partial{\theta}_0} =
\frac{1}{m}
\sum\limits_{i=1}^m (h_0(x^{(i)})-y^{(i)})x^{(i)}_j 
$$
$$
\frac{\partial{J}(\theta)}{\partial{\theta}_0} = \left (
\frac{1}{m}
\sum\limits_{i=1}^m (h_0(x^{(i)})-y^{(i)})x^{(i)}_j \right) +
\frac{\lambda}{m} \theta_j
$$